linter/lints/common.go

```diff
@@ -35,3 +37,39 @@ func GetExtWithOID(exts []pkix.Extension, oid asn1.ObjectIdentifier) *pkix.Exten
 	}
 	return nil
 }
+
+// ReadOptionalASN1BooleanWithTag attempts to read and advance incoming to
+// search for an optional DER-encoded ASN.1 element tagged with the given tag.
+// Unless outExpectedTagPresent or outPresent are nil, it stores whether the tag
+// was found in outExpectedTagPresent and an element with the tag was found in
+// outPresent, otherwise each boolean will take the default value. It reports
+// whether all reads were successful.
+func ReadOptionalASN1BooleanWithTag(incoming *cryptobyte.String, out *bool, tag cryptobyte_asn1.Tag, defaultValue bool) bool {
+	if out != nil {
+		*out = defaultValue
+	}
+
+	// ReadOptionalASN1 performs a peek and will not advance if the tag is
+	// missing, meaning that incoming will retain bytes.
+	var valuePresent bool
+	var valueBytes cryptobyte.String
+	ok := incoming.ReadOptionalASN1(&valueBytes, &valuePresent, tag)
+	if !ok {
+		return false
+	}
+
+	if valuePresent {
+		// X.690 (07/2002) section 8.2 states that a boolean will have length of 1
+		// and value true will have contents FF.
+		// https://www.itu.int/rec/T-REC-X.690-200207-S/en
+		boolBytes := []byte(valueBytes)
+		if len(boolBytes) == 1 && boolBytes[0] == 0xFF {
+			if out != nil {
+				*out = true
+			}
+		}
```

Comment:
Now that we take a `defaultValue`, this code is subtly incorrect. Imagine an ASN.1 spec of `BOOLEAN DEFAULT TRUE`. Those are (vanishingly?) rare but we should still handle it right. This should look something like:

```go
val := defaultValue
if valuePresent {
  if bytes.Equal(boolBytes, []byte{0xFF}) {
    val = true
  } else if bytes.Equal(boolBytes, []byte{0x00}) {
    val = false
  } else {
    // unrecognized encoding of boolean!
    return false
  }
}
if out != nil {
  *out = output
}
```

This handles the case where a field is defined to have a default of TRUE, and we encounter DER-encoded bytes containing the value FALSE (0x00).

Note! This assumes the only valid encodings for boolean are `0x00` (FALSE) and `0xFF` (TRUE). This is a DER-specific requirement (i.e. BER does not require it).

X.690 02/2021:
> 11.1 Boolean values
> If the encoding represents the boolean value TRUE, its single contents octet shall have all eight
> bits set to one. (Contrast with 8.2.2.)

> 8.2.2 If the boolean value is:
>  FALSE
> the octet shall be zero.
> If the boolean value is
>  TRUE
> the octet shall have _any non-zero value_, as a sender's option.

(emphasis mine)

In other words, the basic (BER) encoding says any nonzero is TRUE. The distinguished (DER) encoding locks it down a little more and says only 0xFF is TRUE. Anything else: 0x01, 0x23, 0xFE, is invalid encoding.